### PR TITLE
Improve Typing: add typescript support for deferrable, new Primary/Unique syntax

### LINKED ISSUE #3944: Inconsistent camelCase use for datetime in schema
# Environment

Knex version: 0.21.2

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

The docs describe creating a `datetime` column by using `table.datetime()`: http://knexjs.org/#Schema-datetime

However, the TypeScript definition and tests seem to indicate it should be `table.dateTime()`:

https://github.com/knex/knex/blob/b59161f9944acf858b149a6c2901b14fcd4ba8aa/types/index.d.ts#L1560

https://github.com/knex/knex/blob/b59161f9944acf858b149a6c2901b14fcd4ba8aa/test/integration/builder/selects.js#L743

as a matter of fact, I was only able to find a single reference that seems to use `datetime` in schema building:

https://github.com/knex/knex/blob/b59161f9944acf858b149a6c2901b14fcd4ba8aa/test/unit/schema/mysql.js#L832

Are the docs wrong, or is there something bigger here I'm missing? (Also, how does that last test pass when using `datetime()` instead of `dateTime()`? Are both valid? @lorefnon

### PR DESCRIPTION
**Pull request by zeotuan**
also addressing missing datetime missing in 
close #3944

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -1908,12 +1909,16 @@ export declare namespace Knex {
     uuid(columnName: string): ColumnBuilder;
     comment(val: string): TableBuilder;
     specificType(columnName: string, type: string): ColumnBuilder;
+    primary(columnNames: readonly string[], options?: Readonly<{constraintName?: string, deferrable?: deferrableType}>): TableBuilder;
```
Comment (zeotuan) [inline types/index.d.ts:1912]: New type for primary using options object

File: types/index.d.ts [hunk 2]
```diff
@@ -1908,12 +1909,16 @@ export declare namespace Knex {
...
+    /** @deprecated */
     primary(columnNames: readonly string[], constraintName?: string): TableBuilder;
     index(
       columnNames: string | readonly (string | Raw)[],
       indexName?: string,
       indexType?: string
     ): TableBuilder;
+    unique(columnNames: readonly (string | Raw)[], options?: Readonly<{indexName?: string, deferrable?: deferrableType}>): TableBuilder;
```
Comment (zeotuan) [inline types/index.d.ts:1920]: Added type for unique using options object

File: types/index.d.ts [hunk 3]
```diff
@@ -1908,12 +1909,16 @@ export declare namespace Knex {
     uuid(columnName: string): ColumnBuilder;
     comment(val: string): TableBuilder;
     specificType(columnName: string, type: string): ColumnBuilder;
+    primary(columnNames: readonly string[], options?: Readonly<{constraintName?: string, deferrable?: deferrableType}>): TableBuilder;
+    /** @deprecated */
```
Comment (zeotuan) [inline types/index.d.ts:1913]: add deprecate tag for old syntax

**Comment by zeotuan** [comment]
@kibertoad Hi i added type support for the the new deferrable support that was added. Also added a datetime type

File: types/index.d.ts [hunk 4]
```diff
@@ -1884,6 +1884,7 @@ export declare namespace Knex {
     boolean(columnName: string): ColumnBuilder;
     date(columnName: string): ColumnBuilder;
     dateTime(columnName: string, options?: Readonly<{useTz?: boolean, precision?: number}>): ColumnBuilder;
+    datetime(columnName: string, options?: Readonly<{useTz?: boolean, precision?: number}>): ColumnBuilder;
```
Comment (kibertoad) [inline types/index.d.ts:1887]: what is it for? isn't this overload fully equivalent to the previous one?

Comment (zeotuan) [reply types/index.d.ts:1887]: I think that the problem was mentioned in #3944 where the same method is aliased as ```dateTime```and ```datetime``` but only the definition for ```dateTime``` is available in TS defs while only the ```datetime``` is mentioned in documentation so i added ```datetime``` to the TS defs.

Comment (kibertoad) [reply types/index.d.ts:1887]: ah, gotcha, I missed the difference in casing. thank you

**Review by kibertoad — APPROVED** [review]

**Pull request merged**